packages/lit-html/src/static.ts

```diff
@@ -0,0 +1,109 @@
+/**
+ * @license
+ * Copyright (c) 2020 The Polymer Project Authors. All rights reserved.
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * The complete set of authors may be found at
+ * http://polymer.github.io/AUTHORS.txt
+ * The complete set of contributors may be found at
+ * http://polymer.github.io/CONTRIBUTORS.txt
+ * Code distributed by Google as part of the polymer project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+import {html as coreHtml, svg as coreSvg, TemplateResult} from './lit-html.js';
+
+/**
+ * Wraps a string so that it behaves like part of the static template
+ * strings instead of a dynamic value.
+ *
+ * This is a very unsafe operation and may break templates if changes
+ * the structure of a template. Do not pass user input to this function
+ * without sanitizing it.
+ *
+ * Static values can be changed, but they will cause a complete re-render
+ * since they effectively create a new template.
+ */
+export const unsafeStatic = (value: string) => ({
+  _$litStatic$: value,
+});
+
+type StaticValue = ReturnType<typeof unsafeStatic>;
+
+const stringsCache = new Map<string, TemplateStringsArray>();
+
+/**
+ * Generates a template literal tag function that returns a TemplateResult with
+ * the given result type.
+ */
+const tag = (coreTag: typeof coreHtml) => (
+  strings: TemplateStringsArray,
+  ...values: unknown[]
+): TemplateResult => {
+  let key = strings[0];
```

Comment:
Is it worth having the `stringsCache` at all? We create the key string for every call which iterates and joins everything into a string. The cache is used to avoid iterating and making an array with combined static strings. Those processes seem pretty similar and therefore it doesn't seem worth avoiding the array creation. Plus, there's extra work to fill the cache. Maybe benchmark it?

Reply:
You need a cache of the strings array, since you need to return _the same_ strings array for every call with the same strings+statics (key).

Reply:
It does seem like this can all be done with one loop plus a join as below, though. Are you creating the key first via concat (rather than join) to bias the cache hit to be faster, assuming that concat is faster than creating the strings array and joining it?

Example input:
```js
html`string1${unsafeStatic('static1')}string2${value1}string3${unsafeStatic('static2')}${value2}string4`
```
Do a single loop over `strings`/`values` and create the collapsed array of strings + non-static values, resulting in:
```js
strings = ['string1static1string2', 'string3static2', 'string4'];
values = [value1, value2];
```

Join strings to get key, and use the strings we just made if it doesn't exist yet:
```js
const key = strings.join('$lit$part$');
let cachedStrings = stringsCache.get(key);
if (!cachedStrings) {
  stringsCache.set(key, cachedStrings = strings
}
return coreTag(cachedStrings, values);
```

Reply:
Either way, the non-static values could be accumulated in the first loop creating the key, rather than running a separate filter at the end.

Reply:
String concat is a bit faster, but it might not be worth the code

Reply:
Refactored to one loop